test/e2e/framework.go

```diff
@@ -1145,3 +1172,64 @@ func (data *TestData) mutateAntreaConfigMap(mutatingFunc func(data map[string]st
 	}
 	return nil
 }
+
+// graceExitAntreaController copies Antrea controller binary coverage data file out before terminating the pod
+func (data *TestData) graceExitAntreaController(covDir string) error {
+	rc, stdout, _, err := provider.RunCommandOnNode(masterNodeName(), fmt.Sprintf("kubectl get pods -l=app=antrea,component=antrea-controller -n %s --no-headers=true", antreaNamespace))
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when graceful exit Antrea controller: get antrea-controller pod name, error: %v", err)
+	}
+	podName := strings.Split(stdout, " ")[0]
+	rc, stdout, _, err = provider.RunCommandOnNode(masterNodeName(), fmt.Sprintf("kubectl exec -i %s -n %s -- pgrep %s", podName, antreaNamespace, antreaControllerCovBinary))
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when graceful exit Antrea controller: pgrep %s, error: %v", antreaControllerCovBinary, err)
+	}
+	pid := stdout
+	rc, _, _, err = provider.RunCommandOnNode(masterNodeName(), fmt.Sprintf("kubectl exec -i %s -n %s -- kill -SIGINT %s", podName, antreaNamespace, pid))
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when graceful exit Antrea: send SIGINT signal, error: %v", err)
+	}
+	// a filepath-friendly timestamp format.
+	const timeFormat = "Jan02-15-04-05"
+	timeStamp := time.Now().Format(timeFormat)
+	rc, _, _, err = provider.RunCommandOnNode(masterNodeName(), fmt.Sprintf("kubectl cp %s/%s:%s %s/controller%s.out", antreaNamespace, podName, antreaControllerCovFile, covDir, timeStamp))
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when graceful exit Antrea: copy %s from pod %s, error:%v", antreaControllerCovFile, podName, err)
+	}
+	return nil
```

Comment:
I doubt that you need `kubectl` for most of this? Why not use the K8s APIs. The framework also already has some functions that you should consider using, e.g. `getAntreaController`.

Same comment for `graceExitAntreaAgent` below.

Reply:
thanks, changed to use k8s APIs for some operations.